client/commands/v2/pysa_server.py

```diff
@@ -60,6 +65,77 @@ def __init__(
         self.binary_location = binary_location
         self.server_identifier = server_identifier
 
+    def invalid_model_to_diagnostic(
+        self,
+        invalid_model: query.InvalidModel,
+    ) -> lsp.Diagnostic:
+        return lsp.Diagnostic(
+            range=lsp.Range(
+                start=lsp.Position(
+                    line=invalid_model.line - 1, character=invalid_model.column
+                ),
+                end=lsp.Position(
+                    line=invalid_model.stop_line - 1,
+                    character=invalid_model.stop_column,
+                ),
+            ),
+            message=invalid_model.full_error_message,
+            severity=lsp.DiagnosticSeverity.ERROR,
+            code=None,
+            source="Pysa",
+        )
+
+    def invalid_models_to_diagnostics(
+        self,
+        invalid_models: Sequence[query.InvalidModel],
+    ) -> Dict[Path, List[lsp.Diagnostic]]:
+        result: Dict[Path, List[lsp.Diagnostic]] = {}
+        for model in invalid_models:
+            if model.path is None:
+                self.log_and_show_message_to_client(
+                    "Path cannot be None", lsp.MessageType.WARNING
+                )
+            else:
+                result.setdefault(Path(model.path), []).append(
+                    self.invalid_model_to_diagnostic(model)
+                )
+        return result
+
+    async def update_errors(self, document_path: Path) -> None:
+        await _publish_diagnostics(self.output_channel, document_path, [])
+        pyre_connection = api_connection.PyreConnection(
+            Path(self.pyre_arguments.global_root)
+        )
+        try:
+            model_errors = query.get_invalid_taint_models(pyre_connection)
+            diagnostics = self.invalid_models_to_diagnostics(model_errors)
+            await self.show_model_errors_to_client(diagnostics)
+        except PyreQueryError as e:
+            await self.log_and_show_message_to_client(
+                f"Error querying Pyre: {e}", lsp.MessageType.WARNING
+            )
+
+    async def copy_model(self, document_path: Path, position: lsp.Position) -> None:
+        pyre_connection = api_connection.PyreConnection(
+            Path(self.pyre_arguments.global_root)
+        )
+        try:
+            types = query.types(pyre_connection, [document_path])
+            for type in types[0].types:
+                stop = lsp.Position(
+                    line=type.location["stop"]["line"],
+                    character=type.location["stop"]["column"],
+                )
+                if stop.line > position.lline or (
```

Comment:
Typo?

```suggestion
                if stop.line > position.line or (
```

Reply:
I think this logic is too simple. I think we need to check that `position` is between the `start` and `stop` positions of `type`.

Reply:
Sure, I'll add that! I think I had it this way originally since the results are in order (but if anything changes with the `types` query then you're right that this wouldn't work